### PR TITLE
Make sure that the asset directory on devfs always exist.

### PR DESCRIPTION
**Pull request by chingjun**
Fixes the problem where hot reloading assets only works after a hot
restart.

When the bundle is being uploaded for the first time, the files in the
assets are all skipped, therefore the assets directory are not created
on DevFS. Then flutter_tools calls `setAssetDirectory` on the vm
service, which failed because the directory doesn't exist, and the
`DirectoryAssetBundle` is not valid. So asset_manager only has one
resolver, which is the `ApkAssetProvider` on Android, and never sees
updated asset files.

The problem is only corrected after a hot restart.

This PR fixes this issue by uploading a dummy file on firstBundleUpload
to make sure that the directory is properly created.

---

I can't find any github issues describing this, so here's the repro steps:

1. Launch an app with asset
2. Replace the asset file with a new image, then hot reload.
3. The image on screen should be updated, but isn't.
4. Then hot restart, and the image is updated.
5. Update the image again and hot reload. Now the image on screen is correctly updated.

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Does this only happen on Android? IIRC we have some tests that cover this case but the flutter-tester device is used. 

What path does uploading a file cover that creating a new DevFS does not? Perhaps we should make the setAssetDirectory RPC "do the right thing" instead?

**Comment by chingjun** [comment]
I can actually reproduce this on both iOS and Android. I haven't tested on other platforms, but since the code in question are in cross platform C++ in the engine, I assume it happens on other platforms as well.

> What path does uploading a file cover that creating a new DevFS does not?

It creates the directory `code_cache/appUMVZGY/app/build/flutter_assets` for me. Without this step, this directory doesn't exist after the initial devfs sync.

> Perhaps we should make the setAssetDirectory RPC "do the right thing" instead?

Sounds like a good idea. Maybe something like this: https://github.com/chingjun/engine/commit/1c615549dd7b658e7451a44c0ad687414edb58c6, let me know if you prefer to do this in the engine or in the tools.

Thanks!

**Comment by jonahwilliams** [comment]
uh yeah, the engine seems to lack the nice Directory.createSync(recursive: true) APIs we have in Dart.

It would be worth filling a bug to see if we should change this approach, but I guess I'd rather add some more Dart code than more cpp code. This approach LGTM

File: packages/flutter_tools/lib/src/devfs.dart [hunk 1]
```diff
@@ -631,6 +631,13 @@ class DevFS {
           assetPathsToEvict.add(archivePath);
         }
       });
+
+      // When the bundle is first uploaded, include a dummy file to make sure
+      // that the assets directory on DevFS is created.
+      if (bundleFirstUpload) {
+        final Uri deviceUri = _fileSystem.path.toUri(_fileSystem.path.join(assetDirectory, '.__dummy_flutter_asset'));
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/devfs.dart:638]: nit: hoist the name into a constant

**Review by jonahwilliams — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/devfs.dart (hunk 1 above)
Comment (chingjun) [reply packages/flutter_tools/lib/src/devfs.dart]: Done.

**Comment by chingjun** [comment]
Thanks Jonah. Ya, I searched through the header files and can't find a nice way to create directories recursively, otherwise it really would be cleaner that way :(

**Comment by chingjun** [comment]
Filed #85290 for creating directories recursively in fml.

**Pull request merged**